ports/stm/peripherals/stm32f7/stm32f746xx/clocks.c

```diff
@@ -0,0 +1,67 @@
+
+/*
+ * This file is part of the Micro Python project, http://micropython.org/
+ *
+ * The MIT License (MIT)
+ *
+ * Copyright (c) 2020 Lucian Copeland for Adafruit Industries
+ * Copyright (c) 2020 Mark Olsson <mark@markolsson.se>
+ *
+ * Permission is hereby granted, free of charge, to any person obtaining a copy
+ * of this software and associated documentation files (the "Software"), to deal
+ * in the Software without restriction, including without limitation the rights
+ * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+ * copies of the Software, and to permit persons to whom the Software is
+ * furnished to do so, subject to the following conditions:
+ *
+ * The above copyright notice and this permission notice shall be included in
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/mpconfig.h"
+
+void stm32_peripherals_clocks_init(void) {
+  RCC_ClkInitTypeDef RCC_ClkInitStruct;
+  RCC_OscInitTypeDef RCC_OscInitStruct;
+
+  /* Enable Power Control clock */
+  __HAL_RCC_PWR_CLK_ENABLE();
+
+  /* The voltage scaling allows optimizing the power consumption when the device is
+     clocked below the maximum system frequency, to update the voltage scaling value
+     regarding system frequency refer to product datasheet.  */
+  __HAL_PWR_VOLTAGESCALING_CONFIG(PWR_REGULATOR_VOLTAGE_SCALE1);
+
+  /* Enable HSE Oscillator and activate PLL with HSE as source */
+  RCC_OscInitStruct.OscillatorType = RCC_OSCILLATORTYPE_HSE;
+  RCC_OscInitStruct.HSEState = RCC_HSE_BYPASS;
+  RCC_OscInitStruct.HSIState = RCC_HSI_OFF;
+  RCC_OscInitStruct.PLL.PLLState = RCC_PLL_ON;
+  RCC_OscInitStruct.PLL.PLLSource = RCC_PLLSOURCE_HSE;
+  RCC_OscInitStruct.PLL.PLLM = 8;
```

Comment:
I think we need to move the clock code outside of this, various boards use totally different clock settings....the nucleo_f7 uses the internal Osc, whereas the disco_f7 has an external 25mhz one.

Reply:
In general, clock settings tend to differ on the Family level (F4, F7, H7, etc) but within that the only setting we need to change is PLLM, which is typically set to equal the oscillator frequency/1,000,000. That's the purpose of BOARD_OSC_DIV, so there's a clear reference to the oscillator speed in mpconfigboard.h. 

This section could use the hal-conf macros instead, but that requires more specific HAL familiarity and I've been avoiding doing so until there's a more compelling reason.

Reply:
That's fine if the OSC is internal, but on some boards it's not.

For the F7_DISCO port, I'll do a more in-depth look into this. There are quite a few things I've come across that would need to be more configurable by board.

Reply:
I'm not sure what you mean. We do not use the internal OSC on any board.

Reply:
Sorry, my mistake...but RCC_OscInitStruct.HSEState is different on some boards...

Reply:
We assume all boards have an external oscillator which we match with BOARD_OSC_DIV, leaving the rest of the settings the same so that USB and the peripherals receive a consistent clock speed. You can see an example of a 25MHz oscillator setup in `stm32f411ce-blackpill`

Reply:
Yes, that's fine...but the PLLN and PLLQ need to be different if we use the external SRAM on the disco board, which will be needed in the future to get the display to work. It can't be the max 216mhz speed, but max 200mhz for the external SRAM.

Reply:
Interesting. We don't have support for external SRAM yet, but that's a good heads up. If you want to submit a PR for that I'd be happy to take a look. It'd probably be something that could be handled with a one-time mpconfigboard.h flag. It might be worth checking out the F412 as well and see if it has a similar requirement.

Reply:
I'll resolve this...will make an issue about custom clock settings which we can look at later.

Reply:
Please add BOARD_OSC_DIV for now. I think it'll likely be switched over to HSE_VALUE/1000000 very quickly with Scott's low power code, but it'll at least make things consistent for that change.